[#42923] Move followers to a later date when delay covers days changing from working to non-working

cbliard opened the pull request:
It is part of  [OP#42923](https://community.openproject.org/wp/42923) and a continuation of the initial PR #11205 Add job to reschedule on weekend days changes 

~~The special case to handle is about keeping the same ~gap~ distance between work packages following each other when some working days are added. In this case, the reschedule service has to know what were the working days before they changed, and they must be passed down from the `ApplyWorkingDaysChangeJob` to the `SetScheduleService`.~~

EDIT 2022-09-19: it changed since [OP#44053 Disregard distance (not lag) between related work packages when scheduling FS-related work packages](https://community.openproject.org/wp/44053). The only reason a follower would move forward now is to have a violation of follows/precede relation: having a follower starting before its predecessor ends is forbidden. A follower can never move backwards (unless moved by user).

So the below comment from Parimal is no longer relevant

From [Parimal related comment](https://community.openproject.org/projects/openproject/work_packages/31992/activity#activity-155)
> Let's look at how two WPs with a preceeding/following relationship behave today when the date of the former changes (taking non-working days out of the question completely).
> 
> Imagine that the gap between A and B (where B is following A) is 4 days. When you move A back 2 days, the start date of B also moves backwards by two days. However, when you move A forward 2 days, the start date of B does not change; instead, the gap is reduced to 2 days.
> 
> This behaviour might be odd, but the Duration/Non-working days feature do not intend to change it.
> 
> In light of this, let's introduce non-workings days back into the equation. In your scenario, wp2 and wp1 had a gap of 0 working days. When Wednesday is added as working day and we don't affect the dates of wp2, the gap increases to 1 day. Since this increase in gap gap is not manually set (this can be done by manually moving wp2 into the future), the change of Wednesday as a working day should not automatically do this.
> 
> Instead, the start date of wp2 should move back by one day to Wednesday, and the finish date to Thursday (conserving duration, naturally).
> 
> Same behaviour for that second case; if Wednesday is added as a working day, the 4 day gap is retained and wp2 now starts on Wednesday next week (and ends on Thursday).
> 

There is still a case to handle: follower work packages with a follows relation with a delay (also known as lag) will have to move forward if that delay is covering a day changing from working to non-working.

I prefer having this second part in a separate pull request so it can be paused or cancelled any time while still having the first part reviewed, validated and merged.

cbliard committed:
Add acceptance specs

WorkPackages::ApplyWorkingDaysChangeJob
  when a follower has a predecessor with a non-working day between them that is now a working day
    does not move the follower

lindenthal commented:
Hi @cbliard and @psatyal,
I just found this topic by chance. Not sure if I understand the scenario right. Did you check how other tools handle this? What do you mean by "gap"?
Best
Niels

psatyal commented:
Hi @lindenthal,

Did you read the [original discussion in the work package](https://community.openproject.org/projects/openproject/work_packages/31992#activity-154)? This should give you some background.

It's normal that you don't have the full context; you were away for the past two weeks and we've been exchanging with @oliverguenther and @cbliard over questions like these. We always update the original specs once a decision is made. 

The gap refers to a value called "delay" (that exists in the API but is not modifiable directly on the front-end); this is the number of days between two related work package. Imagine WP1 ends on 6 Sept. and the WP2, which follows it, starts on 9 Sept. This is currently possible, and the delay can be considered 2 (if it were zero, WP2 would start the day after, ont 7 Sep). What happens to the start/finish dates of WP2 now depends on whether WP1 days are moved forward (WP2 not affect) or back (WP2 affected).

Christophe's question was essentially what to do with this gap when adding/removing non-working days via admin settings, since we decided that this should affect scheduling. I proposed that we don't change the current behaviour. If you have quesitons or would like clarification, let's discuss it over a call. 

Cheers
Parimal

lindenthal commented:
Hi @psatyal,

Thanks for the explanation. In MS Project this is called "Lag". This is why I was a little confused. @cbliard How much effort is it to rename it to "Lag"?

![image](https://user-images.githubusercontent.com/1019287/188580695-9c46d28e-39ce-4a58-b3e1-8d2e5f7b43bb.png)

I am still not sure if I understand the consequences 100 %. So let's talk about this later. I need to better understand how lag currently works. In general I think it is very problematic to use the lag as scheduling parameter if is not possible to see and edit it. This will definitely create a lot of confusion.

So my next task is to better understand the current implementation of "lag". 
By the way: The lag feature is a feature request by Autoliv
cc @ulferts 

Best
Niels

psatyal commented:
Hi @lindenthal,

Yes, let's discuss this over a call. 

So you have more context, here is a summary of the current behaviour (I use 'lag' instead of 'delay), assuming wp1 is followed by wp2:

- By default, wp2 starts a day after wp1 (lag: 0). Moving wp1 forwards and backwards in time will push/pull wp2 to maintain this lag, and wp2 will alway start the day after wp1 finishes.
- But lag can technically be set by going to Gantt view and dragging wp2 into the future. Imagine wp2 is moved 4 days after its original start date, and lag now is 3 days.
- If wp1 is moved 2 days into the past now, the lag is conserved and wp2 is now also moved to 2 days earlier. The lag is still 3 days.
- However, oddly, when wp1 is dragged foward by 1 day, wp2 is **not** pushed forward by 1 day. Instead, wp2's dates are conserved and the space between them is reduced (i.e, lag is reduced). 
- When moving wp1 forward, wp2 is only pushed if wp1's finish date is now later than wp2's start date (and lag becomes zero again).
- Moving wp1 back a few days now will also maintain _this_ new lag of zero, and wp2 starts the day after wp1 again.

This is the current behaviour, unrelated to duration/non-working days. 

This screencast illustrates this behaviour visually:

https://user-images.githubusercontent.com/88370597/188589583-1b2452bb-7e6e-4604-bc01-51f2cef18a88.mov

Cheers
Parimal

psatyal commented:
Extracting this conversation out of this PR to a new bug report on community:
https://community.openproject.org/projects/openproject/work_packages/44053/activity 

@cbliard Please do have a look at that. We can consider this to not be a blocking element for this PR.

cbliard commented:
Sorry, I think there is a misunderstanding. I only have limited connectivity today.

I used the word gap with the meaning of the number of working days between 2 work package following each other. This gap should stay constant whatever happens with setting days as working or non-working. It is the free space between 2 work package where they can be moved. I could have used the word "distance" maybe. This word just helps me describe the acceptance scenarios.

Parimal:
> The gap refers to a value called "delay" 

No, the delay (or lag) is something. It is a property of a follows relation to let some uncompressible space after the predecessor. The follower can't be moved into that space.

Niels:
> How much effort is it to rename it to "Lag"?

None, as the "gap"/"distance" I was referring too was just about giving a name to "the free space between two work packages following each other" to describe acceptance scenarios more easily.

We have something like lag, which is called delay internally and can only be changed via the API.

And I'd like to clarify this difference between gap/distance (let's call it distance) and delay/lag (let's call it lag) by commenting on Parimal's example:

> - By default, wp2 starts a day after wp1 (lag: 0). Moving wp1 forwards and backwards in time will push/pull wp2 to maintain this lag and wp2 will always start the day after wp1 finishes.

Yes! Lag is 0 by default when creating follows relations, and in this example the distance between work packages is 0 too.

> - But lag can technically be set by going to Gantt view and dragging wp2 into the future. Imagine wp2 is moved 4 days after its original start date, and lag now is 3 days.

No, lag is still 0, the distance is now 3.

> - If wp1 is moved 2 days into the past now, the lag is conserved and wp2 is now also moved to 2 days earlier. The lag is still 3 days.

lag is still 0
and yes, distance is conserved when moving predecessor backwards and so it is still 3 days, constraining the follower to move 2 days earlier.

> - However, oddly, when wp1 is dragged foward by 1 day, wp2 is not pushed forward by 1 day. Instead, wp2's dates are conserved and the space between them is reduced (i.e, lag is reduced).

yes, but I'd rather say that distance is reduced. If it was a lag of 3 instead of a distance, then follower would also have to move forward by 1 day (that's the purpose of defining a lag).

> - When moving wp1 forward, wp2 is only pushed if wp1's finish date is now later than wp2's start date (and lag becomes zero again).

yes
and lag was always zero
it's the distance which changed

> - Moving wp1 back a few days now will also maintain this new lag of zero, and wp2 starts the day after wp1 again.

yes
distance does not change when moving backwards, so the distance of zero is maintained.


And regarding changing the working / non working days, we can have this example where two work packages are following each other, and some days become non-working and then become non-working again:

This is the initial situation where *B* follows *A*; each letter in MTWTF represent a day (Monday Tuesday Wednesday Thursday Friday) 
```
days | MTWTF
wp A | XX
wp B |   XX
```
The `X`s represent the days on which the work package is spanning
* *A* is from Monday to Tuesday
* *B* is from Wednesday to Thursday

If Tuesday is defined as a non-working day, we'll get this
```
days | MTWTF
wp A | X.X
wp B |    XX
```

The `.` represents a non-working day between the start and finish dates of a work package. 

*A* finishes one day later, and *B* moves one day later

If we now set Tuesday to a working day back, we'll get this:
```
days | MTWTF
wp A | XX
wp B |   XX
```

It is the initial situation: *A* finishes one day earlier, and *B* moves one day earlier.

Now if we define Thursday as a non-working day, we get this:
```
days | MTWTF
wp A | XX
wp B |   X.X
```

*B* finishes one day later.

If we now set Thursday to a working day back, we'll get this:
```
days | MTWTF
wp A | XX
wp B |   XX
```

It is the initial situation again: *B* finishes one day earlier.

Now if we define Wednesday as a non-working day, we'll get this:
```
days | MTWTF
wp A | XX
wp B |    XX
```

*B* moves one day later.

And now, that's the situation where I asked Parimal the question: what should happen in this situation if Wednesday is set back to a working day? ~~And his reply was that the number of working days between A and B was 0 before setting Wednesday as a working day (so what I named "distance" was 0), and so after setting Wednesday as a working day, the number of days should remain 0.~~

EDIT 2022-09-19: it changed due to https://community.openproject.org/wp/44053. A follower never moves backward anymore, unless explicitly moved by the user.

~~So if we now set Wednesday to a working day back, we'll get this:~~
```
# WRONG
days | MTWTF
wp A | XX
wp B |   XX
```

~~It is the initial situation again: *B* moves one day earlier.~~

~~In all these examples, the distance between A and B was always 0.~~

EDIT 2022-09-19: *B* does not move at all. The distance is not kept. We get this:
```
days | MTWTF
wp A | XX
wp B |    XX
```



I hope this clarifies the *gap* term I used (for which I now prefer *distance* instead)

cbliard commented:
> Extracting this conversation out of this PR to a new bug report on community: https://community.openproject.org/projects/openproject/work_packages/44053/activity
> 
> @cbliard Please do have a look at that. We can consider this to not be a blocking element for this PR.

@psatyal I did not have any connection till then. I think the bug report is mixing lag and distance, as I explained above. Sorry to not have been able to reply earlier.

The behavior described in the bug is exact, that's what is currently implemented. I do not know if it is right or wrong. It has been implemented like this for a long time. @ulferts maybe you can add more details?

cbliard committed:
Make String#squish remove sql comments

Add `using CoreExtensions::SquishSql` to use it.

cbliard committed:
Introduce create_schedule rspec helper

It works like `let_schedule`, but can be used in the rspec test directly instead of needing a context to define the let.

Also refactor `let_schedule` to use `create_schedule` internally, and remove unused features of `let_schedule`.

cbliard committed:
Fix scheduling bug with followers with delay

When a follows relation has a delay, and this delay covers some
non-working days, the soonest start date must be computed by having the
delay spanning only on working days, not on non-working days.

cbliard committed:
Reschedule work packages with delay when working days change

If delay is covering days with changing working status, they are updated
as well.

ulferts commented on app/services/work_packages/shared/all_days.rb:
```diff
@@ -50,8 +50,9 @@ def due_date(start_date, duration)
-        date
+      def soonest_working_day(date, delay: nil)
+        delay ||= 0
+        date + delay if date
```
Nitpicking: it would be easier to read if this line read:
```
date + delay.days if day
```

cbliard replied:
~~I'll also do it for the `+ 1` / `- 1` that we have on other places then.~~

Well no, I won't. It would create lots of unneeded short-lived objects. I'll only do it here for `delay`.

cbliard replied:
Thanks, updated

ulferts commented on spec/models/work_packages/scopes/covering_days_of_week_spec.rb:
```diff
@@ -0,0 +1,128 @@
+        not_covered1 | X       |
+        not_covered2 |   X     |
+        not_covered3 |    XX   |
+      CHART
```
A work package without any date is not included in the setup.

cbliard replied:
Thanks, added

ulferts reviewed (changes requested):
Looking good and readable. 

I only have tiny remarks.
